Improve the mangling of USRs

jrose-apple opened the pull request:
- Allow typealiases to use substitutions.
- Consistently use 'a' as a mangling operator.
- For generic typealiases, include the alias as context for any generic parameters.

Typealiases don't show up in symbol names, which always refer to canonical types, but they are mangled for debug info and for USRs (unique identifiers used by SourceKit), so it's good to get this right.

jrose-apple commented:
@swift-ci Please test

slavapestov commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1228,28 +1223,54 @@ void ASTMangler::appendProtocolName(const ProtocolDecl *protocol) {
   // For generic types, this uses the unbound type.
-  TypeBase *key = decl->getDeclaredType().getPointer();
+  Type key;
+  if (auto *alias = dyn_cast<TypeAliasDecl>(decl)) {
```
Maybe we should add a getUnboundGenericType() to TypeDecl. You already had the same bit of code in an earlier PR today, and I've written it out in a few places also.

jrose-apple replied:
Fine with me. Does it return UnboundGenericType, or just Type to handle the non-generic case?

slavapestov replied:
I think the latter. No need to do that before merging the PR of course.

slavapestov commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1228,28 +1223,54 @@ void ASTMangler::appendProtocolName(const ProtocolDecl *protocol) {
-#define DECL(id, parent) \
-    case DeclKind::id:
-#include "swift/AST/DeclNodes.def"
+  // Always use Clang names for imported Clang declarations.
```
Is this an accident?

jrose-apple replied:
Oops, yes, that's the next patch I'm working on. Thanks.

slavapestov commented on lib/Demangling/Demangler.cpp:
```diff
@@ -382,7 +382,7 @@ NodePointer Demangler::demangleOperator() {
     case 'X': return demangleSpecialType();
     case 'Z': return createWithChild(Node::Kind::Static, popNode(isEntity));
-    case 'a': return demangleTypeAlias();
+    case 'a': return demangleNominalType(Node::Kind::TypeAlias);
```
Maybe rename demangleNominalType()?

jrose-apple replied:
I wish we had a better name for "nominal or typealias" than "generic type".

jrose-apple committed:
Improve the mangling of typealiases.

- Allow them to use substitutions.
- Consistently use 'a' as a mangling operator.
- For generic typealiases, include the alias as context for any generic
  parameters.

Typealiases don't show up in symbol names, which always refer to
canonical types, but they are mangled for debug info and for USRs
(unique identifiers used by SourceKit), so it's good to get this
right.

jrose-apple commented:
About to push fixes for the simple things, but I figure I can let the tests keep going while I'm waiting for more feedback: [Linux](https://ci.swift.org/job/swift-PR-Linux/6791/), [Mac](https://ci.swift.org/job/swift-PR-osx/6814/)

jrose-apple commented:
Both full tests passed, and the only things since then have been renames and removing commented-out code.

@swift-ci Please smoke test

jrose-apple committed:
[Mangler] Verify USR manglings as well.

jrose-apple commented:
Added additional verification, fixed a bug that would have been caught by that verification, and updated ABI.rst.

@swift-ci Please test

jrose-apple committed:
If a function has an error type, mangle it as '(Error) -> Error'.

This lets the mangling preserve the invariant that functions always
structurally have function types. Error types don't show up in mangled
names often anyway, but it can occur when you ask for the USR of a
function with an invalid type.

jrose-apple committed:
[IDE] Don't produce USRs for nameless parameters.

...and /do/ produce them for 'self' parameters in destructors when
they /are/ named. I suspect this was the cause of the original
problem.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@shahmishal, any idea why the failure isn't getting posted back on GitHub?

shahmishal commented:
It should have been 82e76edbb7170b0fce95eed922fb7e50bfea9bc0 not c7f9cb7548c51da7db2d21efe7d28478c8ee8e78.
```
Setting status of c7f9cb7548c51da7db2d21efe7d28478c8ee8e78 to FAILURE with url https://ci.swift.org/job/swift-PR-Linux/6815/ and message: '**Build failed**
```

shahmishal commented:
@swift-ci Please test

jrose-apple committed:
[Mangler] Use '_' to represent nameless entities.

This can show up when trying to generate USRs for a document with
errors in it. This isn't a great answer because the names it generates
aren't unique (there may be more than one nameless entity with the
same type), but it at least generates valid mangled names.

When generating mangled names for purposes other than USRs, nameless
entities are now checked for by an assertion.

jrose-apple committed:
[Mangler] Handle initializer contexts with no associated variable.

I /think/ this can only come up in invalid code like

    class Outer {
      let _ = { class Inner {} }
    }

but we still try to generate USRs for this. Just use the enclosing
context as the current context so that we still generate /some/ valid
mangling.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
:-( The builder is testing the correct commit but GitHub is showing them out of order.

jrose-apple commented:
@swift-ci Please test Linux

jrose-apple commented:
Previous full-test passed. Still waiting on review from a SourceKit person. (Erik and I have been talking through chat.)

@swift-ci Please smoke test

akyrtzi commented:
Changes LGTM.

eeckstein reviewed (approved):
LGTM!
